src/pages/experiments/[experimentId]/data-exploration/components/heatmap/VegaHeatmap.jsx

```diff
@@ -41,14 +41,16 @@ const VegaHeatmap = (props) => {
     return [];
   };
 
-  spec.height = getAdjustedHeight();
-  spec.width = getAdjustedWidth();
-  spec.axes = getAxes();
-
-  spec.data.forEach((datum) => {
-    // eslint-disable-next-line no-param-reassign
-    datum.values = data[datum.name];
-  });
+  const vegaSpec = {
+    ...spec,
+    height: getAdjustedHeight(),
+    width: getAdjustedWidth(),
+    axes: getAxes(),
+    data: spec.data.map((datum) => ({
+      ...datum,
+      values: data[datum.name],
+    })),
+  };
```

Comment:
this was the main problem, we modified the prop directly, which fucked with the rendering somehow.

Reply:
This is a problem because:

* `data` as a prop was never passed on to `Vega`. That is, a modification to `data` will not cause re-render of `<Vega/>`, only `<VegaHeatmap/>`
* modification of `spec` as a prop will not cause a re-render in and of itself either, because props are meant to be immutable. modifying them within a component is really bad.

So, the solution:

* create a copy of the `spec`, name it `vegaSpec`, and construct it from the `spec` and `data` props
* use that `vegaSpec` to render `<Vega/>`. if any of the things in the spec get modified, we will get a new heatmap